example-site/app/Main.hs

```diff
@@ -1,130 +1,111 @@
-{-# LANGUAGE DeriveGeneric #-}
-{-# LANGUAGE GeneralizedNewtypeDeriving #-}
-{-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE NamedFieldPuns #-}
-{-# LANGUAGE RecordWildCards #-}
 {-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
 
 module Main where
 
-import Control.Lens
-import Data.Aeson as A
-import Data.Aeson.Lens
-import Data.Function (on)
-import Data.List (sortBy)
-import Data.Map as M
-import Data.Monoid
-import Data.Set as S
-import qualified Data.Text as T
-import Data.Text.Lens
-import Data.Time
-import Development.Shake
-import Development.Shake.Classes
-import Development.Shake.FilePath
-import GHC.Generics (Generic)
-import Slick
+import           Control.Concurrent
+import           Control.Lens
+import           Control.Monad              (forM_, forever, liftM, mzero, void,
+                                             when)
+import           Data.Aeson                 as A
+import           Data.Aeson.Lens
+import           Data.Function              (on)
+import           Data.List                  as L (intersperse, sortBy, (\\))
+import           Data.Map                   as M
+import           Data.Monoid
+import           Data.Set                   as S
+import qualified Data.Text                  as T
+import           Data.Text.Lens
+import           Data.Time
+import           Development.Shake
+import           Development.Shake.Classes
+import           Development.Shake.FilePath
+import           Development.Shake.Util     (shakeArgsPruneWith)
+import           GHC.Generics               (Generic)
+import           Slick
+import           Slick.Build
+import           Slick.Serve
+import           System.Console.GetOpt
+import           System.Directory
+import           System.Directory.Extra     (listFilesRecursive, removeFile)
+import           System.Environment
+
+import           Builder
+
+--------------------------------------------------------------------------------
+
+-- | List our custom keys that we need to manage additional behaviour
+--
+data Flags =
+    Preview     -- ^ key to run Warp server
+  | Watch       -- ^ watch for changes
+  deriving (Eq)
+
+-- | Specific flags for Shake to allow custom keys required for specific logic
+--
+flags :: forall a. [OptDescr (Either a Flags)]
+flags =
+  [ Option "" ["preview"      ] (NoArg $ Right Preview    ) "running as preview"
+  , Option "" ["watch"        ] (NoArg $ Right Watch      ) "running with watch"
+  ]
+
+-- | Specific build rules for the Shake system
+--   defines workflow to build the webiste
+buildRules :: Foldable t => t Flags -> Rules ()
+buildRules flags = do
+
+  let isPreviewMode = Preview `elem` flags
```

Comment:
Thought we agreed that `--preview` and `--watch` should be a single CLI option that starts different things inside?

Reply:
`watch` removed; `prune` added instead

Reply:
I sent you a message suggesting other way of doing that.

Reply:
Unfortunately, adding `--prune` by hand won't work, looking at the code Shake stores `stdin` reference on their own and adds `OptsDescr Prune` .. to flags and tries to read flagshttps://hackage.haskell.org/package/shake-0.18.3/docs/src/Development.Shake.Util.html#shakeArgsPrune